usr/lib/common/pkey_utils.c

```diff
@@ -1573,35 +1605,65 @@ struct {                          \
     curve_type = get_cpacf_curve_type(pubkey->template);
     switch (curve_type) {
     case curve_p256:
+        if (sig_len != sizeof(param.P256.sig_r) + sizeof(param.P256.sig_s)) {
+            TRACE_ERROR("Signature has an invalid length of %ld bytes.\n", sig_len);
+            ret = CKR_ARGUMENTS_BAD;
+            goto done;
+        }
+        if (ecpoint_len != sizeof(param.P256.pub_x) + sizeof(param.P256.pub_y)) {
+            TRACE_ERROR("Public key has an invalid length of %ld bytes.\n", ecpoint_len);
+            ret = CKR_ATTRIBUTE_VALUE_INVALID;
+            goto done;
+        }
         if (hash_len > sig_len / 2)
             hash_len = sig_len / 2;
         fc = KDSA_ECDSA_VERIFY_P256;
         hash_off = 32 - hash_len;
-        memcpy(param.P256.sig_r, sig, sig_len);
+        memcpy(param.P256.sig_r, sig, MIN(sig_len, 2 * sizeof(param.P256.sig_r)));
         memcpy(param.P256.hash + hash_off, hash, hash_len);
-        memcpy(param.P256.pub_x, ecpoint, ecpoint_len);
+        memcpy(param.P256.pub_x, ecpoint, MIN(ecpoint_len, 2 * sizeof(param.P256.pub_x)));
         break;
     case curve_p384:
+        if (sig_len != sizeof(param.P384.sig_r) + sizeof(param.P384.sig_s)) {
+            TRACE_ERROR("Signature has an invalid length of %ld bytes.\n", sig_len);
+            ret = CKR_ARGUMENTS_BAD;
+            goto done;
+        }
+        if (ecpoint_len != sizeof(param.P384.pub_x) + sizeof(param.P384.pub_y)) {
+            TRACE_ERROR("Public key has an invalid length of %ld bytes.\n", ecpoint_len);
+            ret = CKR_ATTRIBUTE_VALUE_INVALID;
+            goto done;
+        }
         if (hash_len > sig_len / 2)
             hash_len = sig_len / 2;
         fc = KDSA_ECDSA_VERIFY_P384;
         hash_off = 48 - hash_len;
-        memcpy(param.P384.sig_r, sig, sig_len);
+        memcpy(param.P384.sig_r, sig, MIN(sig_len, 2 * sizeof(param.P384.sig_r)));
```

Comment:
so here you assume `sig_r` and `sig_s` are the same length (which is probably true). But in above length check you add sizeof of each of them individually ? I reader wonders why is this done differently ????
I would use either `2 * xxx` everywhere, or add the individual sizes everywhere, but don't mix it.

Same below for `pub_x` and `pub_y`.

Reply:
ok, done